pymeasure/display/widgets/log_widget.py

```diff
@@ -25,38 +25,148 @@
 import logging
 
 from ..log import LogHandler
-from ..Qt import QtWidgets
+from ..Qt import QtWidgets, QtCore, QtGui
 from .tab_widget import TabWidget
 
 log = logging.getLogger(__name__)
 log.addHandler(logging.NullHandler())
 
 
+class HTMLFormatter(logging.Formatter):
+    level_colors = {
+        "DEBUG": "DarkGray",
+        # "INFO": "Black",
+        "WARNING": "DarkOrange",
+        "ERROR": "Red",
+        "CRITICAL": "DarkRed",
+    }
+
+    html_replacements = {
+        "\r\n": "<br>  ",
+        "\n": "<br>  ",
+        "\r": "<br>  ",
+        "  ": "&nbsp;" * 2,
+        "\t": "&nbsp;" * 4,
+    }
+
+    def format(self, record):
+        formatted = super().format(record)
+
+        # Apply color if a level-color is defined
+        if record.levelname in self.level_colors:
+            formatted = f"<font color=\"{self.level_colors[record.levelname]}\">{formatted}</font>"
+
+        # ensure newlines and indents are preserved
+        for replacement in self.html_replacements.items():
+            formatted = formatted.replace(*replacement)
+
+        # Prepend the level as HTML comment
+        formatted = f"<!--{record.levelname}-->{formatted}"
+
+        return formatted
+
+
 class LogWidget(TabWidget, QtWidgets.QWidget):
     """ Widget to display logging information in GUI
 
     It is recommended to include this widget in all subclasses of
     :class:`ManagedWindowBase<pymeasure.display.windows.managed_window.ManagedWindowBase>`
     """
 
-    def __init__(self, name, parent=None):
+    fmt = '%(asctime)s : %(message)s (%(levelname)s)'
+    datefmt = '%m/%d/%Y %I:%M:%S %p'
+
+    tab_widget = None
+    tab_index = None
+
+    blink_qtimer = QtCore.QTimer()
+    blink_color = None
+    blink_state = False
+
+    def __init__(self, name, parent=None, fmt=None, datefmt=None):
+        if fmt is not None:
+            self.fmt = fmt
+        if datefmt is not None:
+            self.datefmt = datefmt
+
         super().__init__(name, parent)
         self._setup_ui()
         self._layout()
 
+        # Setup blinking
+        self.blink_qtimer.timeout.connect(self.blink)
+        self.handler.connect(self.blinking_start)
+
     def _setup_ui(self):
         self.view = QtWidgets.QPlainTextEdit()
         self.view.setReadOnly(True)
         self.handler = LogHandler()
-        self.handler.setFormatter(logging.Formatter(
-            fmt='%(asctime)s : %(message)s (%(levelname)s)',
-            datefmt='%m/%d/%Y %I:%M:%S %p'
+        self.handler.setFormatter(HTMLFormatter(
+            fmt=self.fmt,
+            datefmt=self.datefmt,
         ))
-        self.handler.connect(self.view.appendPlainText)
+        self.handler.connect(self.view.appendHtml)
 
     def _layout(self):
         vbox = QtWidgets.QVBoxLayout(self)
         vbox.setSpacing(0)
 
         vbox.addWidget(self.view)
         self.setLayout(vbox)
+
+    def blink(self):
```

Comment:
I would rename all `blink*` methods prefix them with `_` to indicate by convention that they are internal methods.

Reply:
Yeah, that's a good one actually